File: articles/python-simulation-bioinformatics-scientific-workflows/julia/python_biology_simulation_kernel.jl
# Python biology simulation cross-language kernel in Julia.

function logistic_growth(initial_population, growth_rate, carrying_capacity, dt, steps)
    population = Float64(initial_population)
    trajectory = Float64[]

    for step in 0:steps
        push!(trajectory, population)
        growth = growth_rate * population * (1.0 - population / carrying_capacity)
        population = max(population + dt * growth, 0.0)
    end

    return trajectory
end

function gc_content(sequence)
    seq = uppercase(sequence)
    valid = [base for base in seq if base in ['A', 'C', 'G', 'T']]
    if length(valid) == 0
        return NaN
    end
    return count(base -> base == 'G' || base == 'C', valid) / length(valid)
end

trajectory = logistic_growth(25.0, 0.35, 1000.0, 0.1, 200)
println("final_population=", round(trajectory[end], digits=5))
println("gc_content=", round(gc_content("ATGCGCGTAATTAACCGGTTACCGTAGCTA"), digits=5))

File: articles/python-simulation-bioinformatics-scientific-workflows/fortran/python_biology_simulation_kernel.f90
! Python biology simulation cross-language kernel in Fortran.

program python_biology_simulation_kernel
  implicit none

  integer, parameter :: steps = 200
  real :: population, growth_rate, carrying_capacity, dt, growth
  integer :: step

  population = 25.0
  growth_rate = 0.35
  carrying_capacity = 1000.0
  dt = 0.1

  do step = 0, steps
    growth = growth_rate * population * (1.0 - population / carrying_capacity)
    population = population + dt * growth
    if (population < 0.0) population = 0.0
  end do

  print *, "final_population:", population
end program python_biology_simulation_kernel
